FSW/src/util/hkUtil.cpp

```diff
@@ -0,0 +1,184 @@
+/* hkUtil.cpp handles NS2 payload housekeeping
+ * Usage:
+ *  module functionality
+ *      system state data collection
+ *      temperature control
+ * 
+ * Modules encompassed:
+ *  houskeeping
+ *  thermal control
+ *
+ * Additional files needed for compilation:
+ *  none
+ */
+
+/* - - - - - - Includes - - - - - - */
+// C++ libraries
+
+// Other libraries
+
+// NS2 headers
+#include "../headers/hkUtil.hpp"
+
+
+/* Module Variable Definitions */
+static HousekeepingData hkData[HK_SAMPLES_TO_KEEP] = {}; // array to store housekeeping data
+static int hkIndex = 0;
+
+/* - - - - - - Module Driver Functions - - - - - - */
+
+/* - - - - - - handleHousekeeping - - - - - - *
+ * Usage:
+ *  Handles all housekeeping functions.
+ *  Samples temperature and power data, handles thermal control
+ *  
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void handleHousekeeping()
+{
+    sampleHousekeepingData();
+    setHeater();
+    hkIndex++;
+    if (hkIndex >= HK_SAMPLES_TO_KEEP) { hkIndex = 0; }
+}
+
+/* - - - - - - Module Helper Functions - - - - - - */
+
+/* - - - - - - setHeater - - - - - - *
+ * Usage:
+ *  Toggles heater based on current optics bench temperature
+ *  Turns heater on when optics temp crosses below HEATER_TEMP_LOW
+ *  Turns heater off when optics temp crosses above HEATER_TEMP_HIGH
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void setHeater()
+{
+    if (latestHkSample.opticsTemp >= HEATER_TEMP_HIGH)
+    {
+        digitalWrite(PIN_HEAT, LOW); // turn heater off
+    }
+    else if (latestHkSample.opticsTemp <= HEATER_TEMP_LOW)
+    {
+        digitalWrite(PIN_HEAT, HIGH); // turn heater on
+    }
+}
```

Comment:
do you think it is necessary to implement a thermostat algorithm instead of a direct if-statement? 

Right now if the temp is hovering right around `HEATER_TEMP_LOW`/`HIGH` then we could end up turning the heater on and off in a fast oscillating manner every time the loop is called, which could potentially cause electrical noise

Depending on the frequency that `handleHousekeeping()` is called with though (i,e. if it is ~1Hz) then this probably isn't something we need to worry about.

Reply:
If the temp drops below HEATER_TEMP_LOW the heater is turned on and it will stay on until the temperature increases above HEATER_TEMP_HIGH, then it will turn off and stay off until the temp drops back down. That way we get a slow oscillation between the LOW and HIGH temps. As long as the range between them is not tiny this shouldn't trigger too rapidly.

Reply:
I don't really see a scenario where the range is exactly on the low or high and it triggers a bunch, but it would probably not be a bad thing to trigger a warning if it hits a bunch, I'm not really sure.